manual/src/cmds/runtime-tracing.etex

```diff
@@ -205,6 +234,83 @@ instead, call the program as below to produce the same result:
 OCAML_RUNTIME_EVENTS_START=1 ./example
 \end{verbatim}
 
+\subsubsection{s-runtime-tracing-custom-events}{Tracing custom events}
+
+The following program make use of the \texttt{Runtime_events.User} and
+\texttt{Runtime_events.Type} modules to declare a custom event providing
+\texttt{span} and \texttt{int} values. The \texttt{tag} extensible variant is
+extended with \texttt{CustomSpan} and \texttt{CustomInt}.
+
+\begin{verbatim}
+type Runtime_events.User.tag += CustomSpan | CustomInt
+
+let count_span =
+    Runtime_events.User.register "count.span" CustomSpan
+        Runtime_events.Type.span
+
+let count_value =
+    Runtime_events.User.register "count.value" CustomInt
+        Runtime_events.Type.int
+
+let count () =
+    Runtime_events.User.write count_span Begin;
+    for i = 1 to 5 do
+        Runtime_events.User.write count_value i
+    done;
+    Runtime_events.User.write count_span End
+
+let () =
+    Runtime_events.start ();
+    for _ = 1 to 3 do
+        count ()
+    done
+\end{verbatim}
+
+On the consumer side, one can use the provided event tag and type to match on
+the relevant events.
+
+\begin{verbatim}
+let span_event_handler domain_id ts event value =
+    (* we're only interested in our CustomSpan event *)
+    match Runtime_events.User.tag event, value with
+    | CustomSpan, Runtime_events.Type.Begin -> Printf.printf "> count begin\n"
+    | CustomSpan, End -> Printf.printf "< count end\n"
+    | _ -> ()
```

Comment:
If I understand correctly, there is no enforced relation between the name of the even (`CustomSpan`, `CustomInt`) and the payload it carries: the producer could emit *both* span and int events for `CustomSpan`, and the customer could receive them. The fact that nothing prevents us to have `CustomSpan` as a tag for non-span events may or may not be an API defect depending on how we think of it.

Questions:
- Would it be interesting to enforce that an event declared with a span payload is only ever emitted with a span payload? This could be done for example by having separate extensible variants for int-events and span-events, or by making the extensible tag a GADT indexed over its intended payload.
- Or, on the countrary, is it intentional that a given event tag can be used with several payload types? In this case, should we consider rewriting this example to have a single `Custom` event tag, and use it for both span and int payloads?

Reply:
That's a good question. Tags are currently used to provide additional semantics to a type, so in that sense they should be an extensible GADT to make sure they are properly used. 

The question I'm wondering about is that custom types can also be used to refine semantics. For example if I write a web server and want to log response latencies, should I use the built-in `int` type so that it can be consumed by generic consumers, or use a custom `int` type used specifically for latencies. I see that there is a hierarchy of semantics: my library web request latencies < web request latencies < latencies < number.  Using types and tags we're basically exposing two semantics for the `int` value: a broad one (type) and a precise one (tag). 
As a consequence users might not have a use for tags if they come up with sufficiently precise custom types, and in that case it is a burden to extend the GADT with a tag for each custom type.

Reply:
(I don't really understand the discussion on custom types, but I will try to have another go on my next read of the proposed chapter.)

Reply:
@TheLortex’s point is not very clear to me either, but if I’m not mistaken this a discussion of the (already merged) API. It doesn’t mean that it isn’t important, but it seems to me that it shouldn’t be blocking for the present PR.

Reply:
Well I think we should have respected the guideline of not merging significant new changes without updating the manual at the same time, so that we would be blocking both the API and the documentation PR. People would be more motivated to have a good API and a good documentation that way. Now instead we are in a situation where we want to merge documentation that we don't really understand for a feature that already shipped with a worse API.

Reply:
My apologies for taking so long to come back to this PR.

The type of an event is what data it contains. We have three default ones, unit, span begin/end and int which correspond to an instantaneous event, the beginning of an event, the end of an event and finally the measurement of something at a point in time. There's also custom types, for which you can provide an encoder/decoder. The intention behind having the default ones is that generic consumers (e.g an opentelemetry or logging library) can do something sensible with the ones they know about.

A good example of a custom type might be a `Histogram` type. You don't want to register a new type for each `Histogram` you track in your program, you only do it once and then use that type to create a `User.t` with a name and tag. An example here might be `200_response_times` or something similar. You could imagine having a `Histogram` for each response code timing.

One complexity is that runtime events is designed so you can consume in- or out-of-process. This is why there's a name and a tag. The name is carried with the event onwards but the tag is not. Having the two separate is a bit of a pain, there may be something that could be done there.

The design was essentially that the most common use cases were simple (units, spans and ints) to get most of the benefits but custom types could give you a lot more freedom but at the cost of a bit more complicated usage.

Eio's use of custom events for tracing is a good example of how the API is currently used: https://github.com/ocaml-multicore/eio/pull/635/files

> People would be more motivated to have a good API and a good documentation that way.

There's a danger that if we insist on getting things perfect the first time, it won't bring forth good contributions. It's a balancing act and with runtime events I recall part of the argument of moving it to otherlibs was the expectation that it wasn't a mature design.

Reply:
> There's a danger that if we insist on getting things perfect the first time, it won't bring forth good contributions. It's a balancing act and with runtime events I recall part of the argument of moving it to otherlibs was the expectation that it wasn't a mature design.

In my experience, the act of writing documentation often reveals improvements to the API -- it is a way to test a design, orthogonal to its use to write code. (I would say that usage examples are good at highlighting weak spots and missing features, while documentation is good at suggesting opportunities to simplify concepts or improve consistency.) I didn't mean to suggest that the code should be perfect in an absolute sense, but that it should have benefited from this first round of improvement before it reaches users of an upstream release.